fake-cucumber/javascript/src/FakeStepDefinitions.ts

```diff
@@ -0,0 +1,103 @@
+import { messages } from 'cucumber-messages'
+import {
+  ParameterTypeRegistry
+} from 'cucumber-expressions'
+import uuidv4 from 'uuid/v4'
+
+import Argument from 'cucumber-expressions/dist/src/Argument'
+import CucumberExpression from 'cucumber-expressions/dist/src/CucumberExpression'
+
+class FakeStepDefinition {
+  public id: string
+  public expression: CucumberExpression
+
+  constructor(
+    public status: string,
+    public pattern: string = null
+  ) {
+    this.id = uuidv4()
+    if (pattern === null) {
+      this.pattern = `{}${status}{}`
+    }
+    this.expression = new CucumberExpression(this.pattern, new ParameterTypeRegistry)
+  }
+
+  public asEnvelope(line: number): messages.Envelope {
+    return new messages.Envelope({
+      stepDefinitionConfig: new messages.StepDefinitionConfig({
+        id: this.id,
+        pattern: new messages.StepDefinitionPattern({
+          source: this.pattern,
+        }),
+        location: new messages.SourceReference({
+          uri: `some/javascript/status-${this.status}.js`,
+          location: new messages.Location({
+            line
+          })
+        })
+      })
+    })
+  }
+}
+
+class FakeMatch {
+  public stepMatchArguments: Argument<any>[] = []
+  public stepDefinitionId: string = ""
+
+  public makeStepMatchArgumentMessages(): messages.StepMatchArgument[] {
+    const stepMatchMessages: messages.StepMatchArgument[] = []
+    this.stepMatchArguments.forEach(match => {
+      const message = new messages.StepMatchArgument({
+        group: match.group,
+        parameterTypeName: match.parameterType.name
+      })
+      stepMatchMessages.push(message)
+    })
+
+    return stepMatchMessages
+  }
+}
+
+class FakeStepDefinitions {
+  public stepDefinitions: FakeStepDefinition[] = []
+
+  constructor() {
+    this.generateStepDefinitions()
+  }
+
+  public ids(): string[] {
+    return this.stepDefinitions.map(sd => sd.id)
+  }
+
+  public makeMessages(): messages.Envelope[] {
+    return this.stepDefinitions.map((stepDefinition, index) => {
+      return stepDefinition.asEnvelope(index * 3)
+    })
+  }
+
+  public matchStep(step: string): FakeMatch {
+    const fakeMatch = new FakeMatch()
+
+    this.stepDefinitions.forEach(stepDefinition => {
+      const match = stepDefinition.expression.match(step)
+      if (match !== null) {
+        fakeMatch.stepMatchArguments = match
+        fakeMatch.stepDefinitionId = stepDefinition.id
```

Comment:
Let's avoid setters too. They are a common source of bugs. It's better to make this field `private readonly` and make sure it's instantiated in the constructor.